src/Database/TigerBeetle/Raw/Queue.hs

```diff
@@ -1,56 +1,96 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE LambdaCase #-}
 module Database.TigerBeetle.Raw.Queue where
 
-import Control.Concurrent.STM (atomically)
-import Control.Concurrent.STM.TQueue (TQueue, writeTQueue)
+import Database.TigerBeetle.Internal.FFI.Client
+import Database.TigerBeetle.Internal.FFI.Client.ClusterId (ClusterId)
 import Data.Text (Text)
-import Data.Text.Foreign qualified as T
-import Data.Word (Word8)
 import Foreign.Marshal.Alloc (alloca)
-import Foreign.Marshal.Array (peekArray, withArray)
-import Foreign.Storable
-import Database.TigerBeetle.Internal.FFI
+import qualified Data.Text.Foreign as T
+import Foreign (sizeOf, Storable (poke))
+import Control.Exception (finally)
+import Foreign.Ptr (Ptr)
 
--- | Response type for handling callbacks
-data Response = Response
-  { responsePacket :: TBPacket,
-    responseData :: [Word8],
-    responseStatus :: PacketStatus
-  }
+data WithClientOps =
+  WithClientOps
+    { onInitFailure :: TBInitStatus -> IO ()
+    , onDeinit :: TBClientStatus -> IO ()
+    , useSubmit :: (TBPacket -> IO TBClientStatus) -> IO ()
+    }
 
+data ClientKind = Echo | Standard
+  deriving (Eq, Ord, Show)
+
+withClient
+  :: ClientKind
+  -> ClusterId
+  -> Text
+  -> WithClientOps
+  -> TBCompletionCallback  
+  -> IO ()
+withClient kind clusterId address ops completionCb = 
+  alloca $ \clientPtr ->
+    T.withCString address $ \addressPtr -> do
+      -- FIXME: need to understand how completion context should be initialized
+      let completionContext = 0
+      cb <- makeCompletionCallback completionCb
+      initStatus <- initFn clientPtr clusterId addressPtr (fromIntegral $ sizeOf addressPtr) completionContext cb
+      finally 
+        (runClient clientPtr initStatus)
+        (freeClient clientPtr)
```

Comment:
`finally` is important to guarantee that we de-init the client, especially before we de-allocate the client ptr.